### PR TITLE
fix: maintain ordering of return value witnesses when constructing ABI

### LINKED ISSUE #1174: Return value witnesses being sorted results in incorrect ABI decoding
### Aim

Computation is done incorrectly when I take a input (public) and return it. 

This seems to be a new issue with 0.4.0

### Expected behavior

Perform computation as the code is written

### Bug

Computation seems be done incorrectly. See the below code snippet where I specifically return a field supplied from the input called `rollup_type`
Input = 1
Output = 0x3 ???? (even though I am simply returning this)

Similarly, I get the current `rollup_subtree_height` from the input and increment it. 
Input = 2
Output = 0x83 ????

There are many other cases too that I have not written here for brevity. For example in one case I pass a struct to the input called "constants". And I simply return that in my output struct. Somehow the output constants is very very different!

### To reproduce

Code:
```
use dep::std;
struct ConstantRollupData {
    private_kernel_vk_tree_root: Field,
    public_kernel_vk_tree_root: Field,
    base_rollup_vk_hash: Field,
    merge_rollup_vk_hash: Field,
}
struct BaseOrMergeRollupPublicInputs {
    rollup_type: u1,
    rollup_subtree_height: Field, 
    constants: ConstantRollupData,
    calldata_hash: [u8; 32],
}
struct PreviousRollupData {
    base_or_merge_rollup_public_inputs: BaseOrMergeRollupPublicInputs
}

struct MergeRollupInputs {
    previous_rollup_data_0: PreviousRollupData,
    previous_rollup_data_1: PreviousRollupData,
}

// combine previous two hashes to create a new hash: 
fn compute_calldata_hash(left: BaseOrMergeRollupPublicInputs, right: BaseOrMergeRollupPublicInputs) -> [u8; 32] {
    let mut combined_arr: [u8; 64] = [0; 64];
    for i in 0..32 {
        combined_arr[i] = left.calldata_hash[i];
        combined_arr[i + 32] = right.calldata_hash[i];
    }
    std::hash::sha256(combined_arr)
}

fn main(merge_rollup_inputs: pub MergeRollupInputs) -> pub BaseOrMergeRollupPublicInputs {
    let left: BaseOrMergeRollupPublicInputs = merge_rollup_inputs.previous_rollup_data_0.base_or_merge_rollup_public_inputs;
    let right: BaseOrMergeRollupPublicInputs = merge_rollup_inputs.previous_rollup_data_1.base_or_merge_rollup_public_inputs;

    constrain left.rollup_type == right.rollup_type;
    constrain left.rollup_subtree_height == right.rollup_subtree_height;
    let current_height = left.rollup_subtree_height;

    let new_calldata_hash = compute_calldata_hash(left, right);
    BaseOrMergeRollupPublicInputs {
        rollup_type: left.rollup_type,
        rollup_subtree_height: current_height + 1,
        constants: left.constants,
        calldata_hash: new_calldata_hash,
    }
}
```

Prover.toml:
```
[merge_rollup_inputs.previous_rollup_data_0.base_or_merge_rollup_public_inputs]
calldata_hash = [44, 242, 77, 186, 95, 176, 163, 14, 38, 232, 59, 42, 197, 185, 226, 158, 27, 22, 30, 92, 31, 167, 66, 94, 115, 4, 51, 98, 147, 139, 152, 36]
rollup_subtree_height = "2"
rollup_type = "1"
[merge_rollup_inputs.previous_rollup_data_0.base_or_merge_rollup_public_inputs.constants]
private_kernel_vk_tree_root = 0
public_kernel_vk_tree_root = 0
base_rollup_vk_hash = 0
merge_rollup_vk_hash = 0

[merge_rollup_inputs.previous_rollup_data_1.base_or_merge_rollup_public_inputs]
calldata_hash = [44, 242, 77, 186, 95, 176, 163, 14, 38, 232, 59, 42, 197, 185, 226, 158, 27, 22, 30, 92, 31, 167, 66, 94, 115, 4, 51, 98, 147, 139, 152, 36]
rollup_subtree_height = "2"
rollup_type = "1"
[merge_rollup_inputs.previous_rollup_data_1.base_or_merge_rollup_public_inputs.constants]
private_kernel_vk_tree_root = 0
public_kernel_vk_tree_root = 0
base_rollup_vk_hash = 0
merge_rollup_vk_hash = 0
```

I ran `nargo prove p`

Verifier's return value is:
```
[return]

calldata_hash = [...]
rollup_subtree_height = "0x0000000000000000000000000000000000000000000000000000000000000083"
rollup_type = "0x0000000000000000000000000000000000000000000000000000000000000003"

[return.constants]
base_rollup_vk_hash = "0x0000000000000000000000000000000000000000000000000000000000000069"
merge_rollup_vk_hash = "0x0000000000000000000000000000000000000000000000000000000000000062"
private_kernel_vk_tree_root = "0x0000000000000000000000000000000000000000000000000000000000000093"
public_kernel_vk_tree_root = "0x00000000000000000000000000000000000000000000000000000000000000b1"
```

Note here how 
* `rollup_type`: in Noir I do `rollup_type = 1` yet `return.rollup_type = 0x3`
* `rollup_subtree_height`: i increment this in noir. Input =1. Output = 0x83????
* `constants` struct is also very very different to whatever my input was!

### Installation method

Binary

### Nargo version

nargo 0.4.0 (git version hash: a4b196a248fdef9e80bf8c0551d83b1cf23c4a39, is dirty: false)

### @noir-lang/noir_wasm version

_No response_

### @noir-lang/barretenberg version

_No response_

### @noir-lang/aztec_backend version

_No response_

### Additional context

_No response_

### Submission Checklist

- [ ] Once I hit submit, I will assign this issue to the Project Board with the appropriate tags.

### PR DESCRIPTION
**Pull request by TomAFrench**
# Related issue(s)

Resolves #1174 

# Description

## Summary of changes

This PR switches the evaluator to track the witnesses for the return value in a `Vec<Witness>` rather than a `BTreeSet<Witness>`. While it's correct to store them as a set inside the `Circuit`, we rely on the ordering of these witnesses for proper ABI decoding and so we need to maintain this when we're constructing the ABI.

See #1174 for an example of this causing improper ABI decoding.

## Test additions / changes

# Checklist

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [x] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [x] I have reviewed the changes on GitHub, line by line.
- [x] I have ensured all changes are covered in the description.

## Documentation needs
- [ ] This PR requires documentation updates when merged.

# Additional context

### CODE REVIEW

**Comment by kevaundray** [comment]
Can you also add an example to ensure we don't regress?

**Comment by rahul-kothari** [comment]
how did we regress in the first place? I think this is new to v0.4 - curious how it happened.

**Comment by TomAFrench** [comment]
The issue is the integration tests make use of the "prove then immediately verify" feature so we're not loading the public inputs from `Verifier.toml`. We switched to this to cut down CI times but I can revert that.

**Comment by kevaundray** [comment]
> The issue is the integration tests make use of the "prove then immediately verify" feature so we're not loading the public inputs from `Verifier.toml`. We switched to this to cut down CI times but I can revert that.

Was it a large effect on CI times?

**Comment by TomAFrench** [comment]
It was a pretty sizable percentage afaik, i.e. double digits.

**Comment by TomAFrench** [comment]
I've made the changes to integration tests in #1179 as that dwarfs the changes in this PR.

**Comment by kevaundray** [comment]
> It was a pretty sizable percentage afaik, i.e. double digits.

oof RIP productivity, lets try to get a larger github runner to counteract this

**Review by kevaundray — APPROVED** [review]
LGTM - I think if separating prove and verify add 10 minutes, then we should have the larger runners in place, before merging

**Pull request merged**
